test/framework/src/main/java/org/elasticsearch/readiness/MockReadinessService.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.readiness;
+
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.env.Environment;
+import org.elasticsearch.node.MockNode;
+import org.elasticsearch.node.Node;
+import org.elasticsearch.plugins.Plugin;
+
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.ServerSocket;
+import java.net.SocketAddress;
+import java.net.SocketOption;
+import java.nio.channels.ServerSocketChannel;
+import java.nio.channels.SocketChannel;
+import java.nio.channels.spi.SelectorProvider;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+public class MockReadinessService extends ReadinessService {
+    /**
+     * Marker plugin used by {@link MockNode} to enable {@link MockReadinessService}.
+     */
+    public static class TestPlugin extends Plugin {}
+
+    private static class MockServerSocketChannel extends ServerSocketChannel {
+        private static final Map<String, MockServerSocketChannel> mockedSockets = new HashMap<>();
+
+        static ServerSocketChannel open(String nodeName) {
+            var mockSocket = new MockServerSocketChannel();
+            mockedSockets.put(nodeName, mockSocket);
```

Comment:
We should override close as well, which should remove the socket from the static map? Otherwise the same node name used across different tests could have unexpected results?

Reply:
Correct; not the case for the current tests but it could definitely happen (and it would be very confusing for whoever writes the new test)

Reply:
I cannot remove these in close, as these mocks are used to hold the information of the socket status for that node (much like `ReadinessService.boundAddress()` in the current version of the tests). But you are right, the static map is a problem as node names could clash. I will find a better alternative.

Reply:
I realized that since we have a `MockReadinessService` to rely on now (as opposed to the first draft were I was trying to use mockito) we don't even need the map anymore: the socket (mocked) is there, so we can avoid some boilerplate and be much closer to the original way of testing. Thanks for pointing this out and making me "look harder".